File: 0022-generate-parentheses/0022-generate-parentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(0,0,n*2,"",n,ans);
        return ans;
    }

    private void helper(int oc,int cc,int slength,String soln,int n,List<String> ans) {
        if (slength == 0) {
            ans.add(soln);
            return;
        }

        if (oc>cc && cc<n) {
            helper(oc,cc+1,slength-1,soln + ")",n,ans);
        }

        if (oc<n) {
            helper(oc+1,cc,slength-1,soln+"(",n,ans);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def helper(r,ans,ob,cb,final_ans):
            if r == 0:
                final_ans.append(ans)

                return 
            else:
                if ob>cb and cb<n:
                    helper(r-1,ans+")",ob,cb+1,final_ans)
                if ob<n:
                    helper(r,ans+"(",ob+1,cb,final_ans)
                

        final_ans = []

        helper(n,"",0,0,final_ans)

        return final_ans